### PR TITLE
Add vfio-server

### PR DESCRIPTION
**Pull request by Bolodya1997**
# Description
`vfio-server` nse provides `network service -> { MAC address, VLAN tag }` mapping.
# Configuration
* NSE_NAME - A string value of network service endpoint name (default "vfio-server")
* NSE_BASE_DIR - A base directory to create a unix socker for listening incoming requests (default "./")
* NSE_CONNECT_TO - A Network service Manager connectTo URL (default "unix:///var/lib/networkservicemesh/nsm.io.sock")
* NSE_MAX_TOKEN_LIFETIME - A token lifetime duration (default 24h)
* NSE_SERVICES - A list of supported Network Services in inner format:
    Name@Domain: { addr: MACAddr; vlan: VLANTag; labels: Labels; }
    MACAddr = xx:xx:xx:xx:xx:xx
    Labels = label_1=value_1&label_2=value_2
        - Name - a Network Service name
        - Domain - a Network Service domain (don't confuse it with interdomain domains)
        - MACAddr - a MAC address for the Network Service
        - VLANTag - a VLAN tag for the Network Service
        - labelN=valueN - pairs of labels supported by the Network Service
    - Examples:
        - pingpong@worker.domain: { addr: 0a:55:44:33:22:11 }
            - **pingpong** Network Service
            - **worker.domain** Network Service domain
            - **0a:55:44:33:22:11** MAC address

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,13 +2,16 @@ FROM golang:1.15-buster as go
...
+RUN go get github.com/go-delve/delve/cmd/dlv@v1.5.0
+RUN go get github.com/edwarnicke/dl
+RUN dl https://github.com/spiffe/spire/releases/download/v0.11.1/spire-0.11.1-linux-x86_64-glibc.tar.gz | \
+    tar -xzvf - -C /bin --strip=3 ./spire-0.11.1/bin/spire-server ./spire-0.11.1/bin/spire-agent
 
 FROM go as build
 WORKDIR /build
+COPY go.mod go.sum ./
+COPY ./internal/imports ./internal/imports
+RUN go build ./internal/imports
```
Comment (denis-tingajkin) [inline Dockerfile:14]: Why do we need this?
```suggestion

```

File: Dockerfile [hunk 2]
```diff
@@ -20,4 +23,5 @@ CMD dlv -l :40000 --headless=true --api-version=2 test -test.v ./...
 
 FROM alpine as runtime
 COPY --from=build /bin/app /bin/app
+COPY --from=build /bin/dlv /bin/dlv
```
Comment (denis-tingajkin) [inline Dockerfile:26]: Note: you could use target debug.
```suggestion
```

File: internal/networkservice/mapserver/server.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+
+	"github.com/networkservicemesh/cmd-nse-vfio/internal/config"
+)
+
+type mapServer struct {
```
Comment (denis-tingajkin) [inline internal/networkservice/mapserver/server.go:33]: Why it is not part of sdk-sriov?

File: Dockerfile (hunk 1 above)
Comment (Bolodya1997) [reply Dockerfile:14]: It is used to make subsequent builds faster if we don't change dependencies and change only code. It allows docker to use already downloaded dependencies from cache.

File: Dockerfile (hunk 2 above)
Comment (Bolodya1997) [reply Dockerfile:26]: Yes, but there is an additional option to start debugging ourselves with passed env variable:
https://github.com/networkservicemesh/cmd-nse-vfio/blob/a01453eceee67e38b8e3434dbf0b73044d7a3f37/main.go#L77-L79

File: Dockerfile (hunk 1 above)
Comment (denis-tingajkin) [reply Dockerfile:14]: Ah, I got it, our cmd-template is using the Ed's cache because it has not deps :)

File: Dockerfile (hunk 2 above)
Comment (denis-tingajkin) [reply Dockerfile:26]: I'm not sure that we need dlv in runtiume.

File: internal/config/config.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	MACAddr net.HardwareAddr
+	VLANTag int32
+	Labels  map[string]string
+}
+
+// UnmarshalBinary expects string(bytes) to be in format:
+// Name@Domain: { addr: MACAddr; vlan: VLANTag; labels: Labels; }
+// MACAddr = xx:xx:xx:xx:xx:xx
+// Labels = label_1=value_1&label_2=value_2
+func (s *ServiceConfig) UnmarshalBinary(bytes []byte) (err error) {
```
Comment (haiodo) [inline internal/config/config.go:71]: why not use json here?

**Review by haiodo — APPROVED** [review]

**Pull request merged**